### PR TITLE
Improve error formatting, fix edge-case

### PR DESCRIPTION
**Pull request by michaljurecko**
Jira: https://keboola.atlassian.net/browse/PSGO-200

----------

Tento PR sa tyka pkg `internal/utils/errors`:
- Obsahuje veci nad ramec standardnej Go kniznice, formatovanie chyb, primarne pre CLI vystup.
- Nieco z toho sa postupne dostáva priamo do Go, napr. v [Go 1.20](https://blog.devgenius.io/wrapping-multiple-errors-in-go-1-20-78163ef5fc2c) + je to v planoch pre [Go 2](https://go.googlesource.com/proposal/+/master/design/go2draft.md), no zatial to riesime tymto nasim pkg.
- Napr. go chyba ma vyzerat takto `some error`, tento pkg z toho spravi vetu `Some error.`.
- V kode zbierame vsetky chyby ako `multierror` a vypiseme ich naraz, aby ich user mohol fixnut naraz, vysledne su naformatovane ako viac-urovnovy zoznam s odrazkami.
- Viac info je v [pkg doc](https://github.com/keboola/keboola-as-code/blob/main/internal/pkg/utils/errors/errors.go).

Server s dokumentaciou sa da spustit takto:
https://github.com/keboola/keboola-as-code/blob/michaljurecko-PSGO-200/docs/DEVELOPMENT.md#start-documentation-server

<details>
  <summary>Screenshot</summary>

</details>

-----------------

Pri reseni ineho tasku som zistil, ze v jednej situacii sa chyba nevypise dobre:
image

Pridal som na to tento case unit test a v tomto PR to fixuje + par drobnosti co som este nasiel:
https://github.com/keboola/keboola-as-code/pull/1313/commits/1609984fa5d8fe34bd6f99c2a8b22c65e10a5eee

--------

**Kod som okomentoval po commitoch, aby si mal k dispozícii mensie celky, ktore sa daju jednoduchsie pozriet.**
Tj. odporucam to pozriet po commitoch.

### CODE REVIEW

File: internal/pkg/utils/errors/errors.go [hunk 1]
```diff
@@ -76,7 +76,7 @@
 // Example - a custom multi-line error formatting:
 //
 //	 func (e UnencryptedValueError) WriteError(w errors.Writer, level int, trace errors.StackTrace) {
-//	   w.WritePrefix(e.Error(), trace)
+//	   w.WritePrefix(level, e.Error(), trace)
 //	   w.WriteNewLine()
```
Comment (michaljurecko) [inline internal/pkg/utils/errors/errors.go:80]: V prvom commite su drobne vylepsenia ohladom formatovania chyb, vysledok bude dobre vidiet v dalsich commits, v testoch.

File: internal/pkg/utils/errors/writer.go [hunk 2]
```diff
@@ -14,8 +14,8 @@ const (
 type Writer interface {
 	Write(string)
 	WriteIndent(level int)
-	WriteBullet()
-	WritePrefix(prefix string, trace StackTrace)
+	WriteBullet(level int)
+	WritePrefix(level int, prefix string, trace StackTrace)
```
Comment (michaljurecko) [inline internal/pkg/utils/errors/writer.go:18]: Writer sa pouziva, ked si chces vypisat chybu nejak specialne, spravis to tak, ze typ chyby implementuje metodu `WriterError`.

File: internal/pkg/utils/errors/writer.go [hunk 3]
```diff
@@ -85,21 +85,17 @@ func (w *writer) WriteErrorLevel(level int, err error, trace StackTrace) {
...
-				w.Write(w.formatPrefix(w.formatMessage(v.Error(), trace)))
+				w.WriteBullet(level)
+				w.Write(w.formatPrefix(fmt.Sprintf("%s (%T)", w.formatMessage(v.Error(), trace), err)))
 				w.WriteNewLine()
-
-				// Write wrapped error
-				w.WriteIndent(level)
-				w.WriteBullet()
-				w.Write(fmt.Sprintf("%T >>> ", err))
 				w.WriteErrorLevel(level+1, subErr, nil)
```
Comment (michaljurecko) [inline internal/pkg/utils/errors/writer.go:91]: 1. Pkg `errors` ma moznost `WithUnwrap`, ktora vypise aj vsetkych predkov chyby (hlavne pri debugovani).
Tu som tento vypis drobne upravi, aby bolo jasne, ku ktorej chybe sa typ viaze (vid testy s komentarom `1.`).

File: internal/pkg/utils/errors/writer.go [hunk 4]
```diff
@@ -127,48 +123,47 @@ func (w *writer) WriteNestedError(level int, main error, errs []error, trace Sta
...
-			w.WriteErrorsList(level, errs)
-		}
-	} else {
+	if errsCount <= 1 && len(mainStr)+len(subErrsStr) <= 60 && !strings.Contains(subErrsStr, "\n") {
 		w.Write(" ")
-		w.WriteErrorsList(level, errs)
+		w.WriteError(errs[0])
+	} else {
+		w.WriteNewLine()
+		w.WriteErrorsList(level+1, errs)
```
Comment (michaljurecko) [inline internal/pkg/utils/errors/writer.go:137]: 2. Ostatne zmeny su fix pre edge case, kde to vygenerovalo `- - error1` namiesto `- error 1`, vid test.
`level` urcuje, ako velmi je chyba vnorena -> ako ma byt odsadena.

File: internal/pkg/utils/errors/format_test.go [hunk 5]
```diff
@@ -49,6 +74,16 @@ func TestMultiError_Format(t *testing.T) {
...
+func TestMultiErrorGetter_Format(t *testing.T) {
+	t.Parallel()
+	expected := `
+main error:
+- error 1
+- error 2
+`
+	assert.Equal(t, strings.TrimSpace(expected), (multiErrsGetterError{}).Error())
+}
+
```
Comment (michaljurecko) [inline internal/pkg/utils/errors/format_test.go:86]: 2. Doplneny unit test, jedna sa o specificky pripad ked ma chyba implementujuca interface `multiErrorGetter` v sebe dalsi `multierror`.

File: internal/pkg/filesystem/fileloader/fileloader_test.go [hunk 6]
```diff
@@ -281,7 +281,8 @@ func (*testCases) TestFileLoader_ReadYamlFile_Invalid(t *testing.T, fs filesyste
 	assert.Error(t, err)
 	assert.Nil(t, file)
 	expectedError := `
-file "file.yaml" is invalid: yaml: did not find expected key
+file "file.yaml" is invalid:
+- yaml: did not find expected key
```
Comment (michaljurecko) [inline internal/pkg/filesystem/fileloader/fileloader_test.go:285]: 2. V ostatnych testoch sa tym fixlo odsadenie, ktore bolo pri `level=1` chybne o 4 znaky, namiesto 2.

File: internal/pkg/utils/errors/format_test.go [hunk 7]
```diff
@@ -113,9 +113,9 @@ func TestMultiError_FormatWithUnwrap(t *testing.T) {
 	expected := `
 - error 1
 - error with debug trace
-- wrapped2: wrapped1: error 2:
-  - *fmt.wrapError >>> wrapped1: error 2:
-    - *fmt.wrapError >>> error 2
+- wrapped2: wrapped1: error 2 (*fmt.wrapError):
+  - wrapped1: error 2 (*fmt.wrapError):
+    - error 2
 - my prefix:
```
Comment (michaljurecko) [inline internal/pkg/utils/errors/format_test.go:119]: 1. A tu je uprava vypisovania chyb pri zapnuti `WithUnwrap`, predtym nebolo jasne ku ktorej chybe sa dany typ viaze.

File: internal/pkg/utils/errors/format_test.go [hunk 8]
```diff
@@ -137,53 +137,53 @@ func TestMultiError_FormatWithUnwrap(t *testing.T) {
...
-    - z [%s/errors_test.go:34]
-  - sub3 with format [%s/errors_test.go:40]:
-    - this is a very long line from error message, it is printed on new line [%s/errors_test.go:36]
-  - sub4 [%s/errors_test.go:41]:
-    - 1 [%s/errors_test.go:42]
-    - 2 [%s/errors_test.go:43]
-    - 3 [%s/errors_test.go:44]
-- last error [%s/errors_test.go:49]
+- error 1 [%s/errors_test.go:%d]
+- error with debug trace [%s/errors_test.go:%d]
```
Comment (michaljurecko) [inline internal/pkg/utils/errors/format_test.go:141]: V `[]` sa vypisuje miesto vzniku chyby, ak je pri formate zapnute `WithStack`, kedze sa test menil, tak cisla riadkov uz nesedeli, tak som ich nahradil za `%d`.

File: internal/pkg/utils/errors/format_test.go [hunk 9]
```diff
@@ -426,13 +426,11 @@ func (e customError) WriteError(w Writer, level int, _ StackTrace) {
 	w.WritePrefix(level, fmt.Sprintf("this is a custom error message (%s)", e.error.Error()), nil)
 	w.WriteNewLine()
 
-	w.WriteIndent(level)
-	w.WriteBullet()
+	w.WriteBullet(level + 1)
```
Comment (michaljurecko) [inline internal/pkg/utils/errors/format_test.go:429]: `WriteBullet` uz interne vola `WithIndent`, .... nema zmysel volat to bez toho.

File: internal/pkg/encoding/json/schema/error.go [hunk 10]
```diff
@@ -31,8 +31,9 @@ func (v ValidationError) Error() string {
 	return v.message
 }
 
-func (v ValidationError) WriteError(w errors.Writer, _ int, _ errors.StackTrace) {
+func (v ValidationError) WriteError(w errors.Writer, level int, _ errors.StackTrace) {
 	// Disable other formatting
+	w.WriteBullet(level)
 	w.Write(v.Error())
 }
```
Comment (michaljurecko) [inline internal/pkg/encoding/json/schema/error.go:38]: `WriteError` metoda je uz zodpovedna za vypis celej chyby, predtym to bolo divne, ze odrazku vypisala libka a metoda doplnila zvysok - co bolo divne hlavne ak si chcel vypisat viacero riadkov.

Pri prvom si nemusel vypisat `WriteBullet` ale pri dalsich ano.

File: internal/pkg/plan/encrypt/plan.go [hunk 11]
```diff
@@ -67,18 +67,16 @@ func (v UnencryptedValueError) Error() string {
 	return fmt.Sprintf(`%s "%s" contains unencrypted values`, v.kind, v.filePath)
 }
 
-// WriteError prints an example of the found files to the output.
+// WriteError writes formatted bullet list.
```
Comment (michaljurecko) [inline internal/pkg/plan/encrypt/plan.go:70]: Typo

File: docs/DEVELOPMENT.md [hunk 12]
```diff
@@ -10,54 +12,70 @@ cd keboola-as-code
...
 
+### Start Documentation Server
+
+To start the [Go Doc](https://go.dev/doc/) documentation server, run the command bellow, then open http://localhost:6060/pkg/github.com/keboola/keboola-as-code/?m=all:
+```
+make godoc
+```
+
+`m=all` is important to show also private packages.
+
```
Comment (michaljurecko) [inline docs/DEVELOPMENT.md:78]: Do dev `quick start` som pridal, ako si spustit `godoc` server.

**Review by zajca — APPROVED** [review]
Sorry včera sem to nedokoukal, chápu, super PR takto okomentovaný.Škoda že to musíš teda "zatím" implementovat takový handling.

**Comment by michaljurecko** [comment]
Dik. Uplne staci pozriet PR na dalsi den (ak by bolo nieco potrebne ASAP napisem).

-----

Na `multierror` je "industry standard" pouzivat https://github.com/hashicorp/go-multierror, lenze nam to uplne nestacilo a nechceli sme udrziavat vlastny fork.
